revocation/internal/crl/crl.go

```diff
@@ -0,0 +1,298 @@
+				// permanently revoked
+				return &result.CRLResult{
+					Result:         result.ResultRevoked,
+					ReasonCode:     result.CRLReasonCode(revocationEntry.ReasonCode),
+					RevocationTime: revocationEntry.RevocationTime,
+					URI:            crlURL,
+				}, nil
+			}
+		}
+	}
+
+	// check if the revocation with CertificateHold or RemoveFromCRL
+	if len(tempRevokedEntries) > 0 {
+		// sort by revocation time (ascending order)
+		sort.Slice(tempRevokedEntries, func(i, j int) bool {
+			return tempRevokedEntries[i].RevocationTime.Before(tempRevokedEntries[j].RevocationTime)
+		})
+
+		// the revocation status depends on the most recent one
+		lastEntry := tempRevokedEntries[len(tempRevokedEntries)-1]
+		if int(result.CRLReasonCodeRemoveFromCRL) != lastEntry.ReasonCode {
+			return &result.CRLResult{
+				Result:         result.ResultRevoked,
+				ReasonCode:     result.CRLReasonCode(lastEntry.ReasonCode),
+				RevocationTime: lastEntry.RevocationTime,
+				URI:            crlURL,
+			}, nil
+		}
+	}
+
+	return &result.CRLResult{
+		Result: result.ResultOK,
+		URI:    crlURL,
+	}, nil
+}
+
+type entryExtensions struct {
+	// invalidityDate is the date when the key is invalid.
+	invalidityDate time.Time
+}
+
+func parseEntryExtensions(entry x509.RevocationListEntry) (entryExtensions, error) {
+	var extensions entryExtensions
+	for _, ext := range entry.Extensions {
+		switch {
+		case ext.Id.Equal(oidInvalidityDate):
+			var invalidityDate time.Time
+			rest, err := asn1.UnmarshalWithParams(ext.Value, &invalidityDate, "generalized")
+			if err != nil {
+				return entryExtensions{}, fmt.Errorf("failed to parse invalidity date: %w", err)
+			}
+			if len(rest) > 0 {
+				return entryExtensions{}, fmt.Errorf("invalid invalidity date extension: trailing data")
+			}
+
+			extensions.invalidityDate = invalidityDate
+		default:
+			if ext.Critical {
+				// unsupported critical extensions is not allowed. (See RFC 5280, Section 5.2)
+				return entryExtensions{}, fmt.Errorf("CRL entry contains unsupported critical extension: %v", ext.Id)
+			}
+		}
+	}
+
+	return extensions, nil
+}
+
+func download(ctx context.Context, crlURL string, client *http.Client) (*x509.RevocationList, error) {
+	// validate URL
+	parsedURL, err := url.Parse(crlURL)
+	if err != nil {
+		return nil, fmt.Errorf("invalid CRL URL: %w", err)
+	}
+	if strings.ToLower(parsedURL.Scheme) != "http" {
+		return nil, fmt.Errorf("unsupported scheme: %s. Only supports CRL URL in HTTP protocol", parsedURL.Scheme)
+	}
+
+	// download CRL
+	req, err := http.NewRequestWithContext(ctx, http.MethodGet, crlURL, nil)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create CRL request: %w", err)
+	}
+	resp, err := client.Do(req)
+	if err != nil {
+		return nil, fmt.Errorf("request failed: %w", err)
+	}
+	defer resp.Body.Close()
+
+	// check response
+	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
+		return nil, fmt.Errorf("failed to download with status code: %d", resp.StatusCode)
```

Comment:
Please add the `url` into the error message, so user gets to know which url failed the download.

Reply:
Updated.